doc/developer/design/20211126-microbenchmarks.md

```diff
@@ -0,0 +1,135 @@
+# Microbenchmarking framework
+
+## Summary
+
+In order to characterize the performance of the product and guard against regressions, we need to be able to benchmarks the invidiual operations Mz performs when it comes to ingesting and processing data and materializing views.
+
+Microbenchmarking will be in addition to any end-to-end benchmarks that exist or will be developed. It aims to provide a more focused spotlight on regressions: instead of a 2% regression in an entire industry benchmark between product versions X and Y, we will be looking at a 20% regression in a specific named operation between specific commits M and N.
+
+## Goals
+
+Measurements:
+- measure the performance of the query optimization stage, to avoid the introduction of O(N^M) algorithms in the optimization pipeline
+- measure the performance of ingestion of various sources, envelope types and data shapes
+- measure the performance of ad-hoc SELECT queries, both those that construct a dataflow and those do not
+- measure the ingestion performance of materialized operators, that is, how fast a materialized query returns an up-to-date result
+
+Output:
+- provide a report that compares the branch under development/test to some benchmark, such as `main` or `materialized/latest`.
+
+Usability goals:
+- allow for individual microbenchmarks to be specified intuitively
+- be runnable on a laptop while remaining compatible with the CI
+- do not require setting any knobs, thresholds, etc.
+
+## Non-Goals
+
+- produce graphs or any other collateral for marketing purposes
+- benchmark end-to-end customer workloads or run industry-standard benchmarks
+- benchmark scaling, concurrency (except possibly concurrent ingestion of a single stream)
+- provide any sort of "transactions-per-second"(and "per-dollar") values of any kind
+
+## Description
+
+### Specifying individual tests
+
+An individual tests is a python class that specifies the actions that need to be taken prior to executing the operation under test, the thing to be tested, and any cleanup.
+
+For example, this microbenchmark will create a Kafka topic, ingest 100000 rows into it and measure the time it takes for the source to catch up.
+
+```
+class PartitionedKafkaIngestion(WallclockUntilSucces):
+   COUNT = 100000
+   def benchmark(c: Context):
+       return Benchmark(
+           setup = [
+               KafkaCreateTopic(partitions=5),
+               KafkaIngest('{"f1": 1} {"f2": 1}'),
+               CreateSource()
+           ],
+           benchmark = Sql("SELECT COUNT(*) = {COUNT} FROM {SOURCE}")
+       )
+```
+
+Internally, each of the helper classes, such as `KafkaCreateTopic` will produce a testdrive fragment such as `$ kafka-create-topic partitions=5` that will be passed on to testdrive for execution.
```

Comment:
This seems like rather a lot of work! Is there a benefit to the Python wrapper classes over just supplying a testdrive setup script? (Possibly one that’s inline in the Python file.)

To be honest I’d almost be more excited if we reimplemented the necessary logic in Python. I’ve gotten increasingly tired of having to duplicate logic available in the Confluent Python libraries in testdrive because there is no Rust equivalent. Some of the testdrive scripts would be much less tortured if they were just a Python scripts using normal Python libraries.

Reply:
> This seems like rather a lot of work! Is there a benefit to the Python wrapper classes over just supplying a testdrive setup script? (Possibly one that’s inline in the Python file.)

An inline script would work, yes.

The benefits to express in python would be:
- making the feature benchmark benchmarks more concise with less syntactic overhead, code duplication
- make sure that each testdrive is produced fully formed, with `create-topic`, a comment specifying the name of the benchmark it came from and whatnot.
- be able to parameterize each test so that if you ingest `N` values in `kafka-ingest`, your `SELECT COUNT(*)` query later also expects `N` values

> To be honest I’d almost be more excited if we reimplemented the necessary logic in Python. I’ve gotten increasingly tired of having to duplicate logic available in the Confluent Python libraries in testdrive because there is no Rust equivalent. Some of the testdrive scripts would be much less tortured if they were just a Python scripts using normal Python libraries.

That would totally be worth it, I agree!! Right now we have .td fragments attached to various `mzworkflows.py` files, that would not have to be there if the entire test was specified end-to-end 100% in python. This has bothered me.

For this particular project however:
- I would like to offload as much work as possible (e.g. json parsing, retrying, etc.) to existing code in testdrive so that I can be able to quickly get to tackling the elephant in the room for this project, which is producing stable numbers across executions without running each individual feature benchmark 1M times. 
- At a previous company, python proved a bad vehicle for sending benchmark work to the database to do -- it was pegged at 100% for uncomfortable stretches. Granted, there the problem was that the database driver was pure-python, but it kind of drove home the point that python can have disproportionate overhead for operations one thinks as fairly simple byte-pushing. (Calculating simple stuff such as standard deviation using pure-python libraries was also stupidly slow for even small array sizes.)